### ISSUE TITLE
Unclear what "consul.runtime.sys_bytes" and "consul.runtime.alloc_bytes" mean

### ISSUE THREAD

**Issue by jrab89** [hashicorp/consul#4734]
#### Overview of the Issue

https://www.consul.io/docs/agent/telemetry.html#memory-usage says that `consul.runtime.sys_bytes` is "the total number of bytes of memory obtained from the OS". I was thinking this might be the virtual memory used by the Consul process, but it's not. It would be helpful to have more detail about what `consul.runtime.sys_bytes` and `consul.runtime.alloc_bytes` actually mean in the telemetry documentation.

#### Reproduction Steps

```
$ sudo top -b -n 1 -p `pgrep consul`
top - 16:19:44 up 24 min,  1 user,  load average: 0.00, 0.01, 0.05
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.0 us,  0.0 sy,  0.0 ni,100.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem :  1015500 total,   447508 free,    95064 used,   472928 buff/cache
KiB Swap:  2097148 total,  2097148 free,        0 used.   741792 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S %CPU %MEM     TIME+ COMMAND
 3936 root      20   0   41044  22768  11360 S  0.0  2.2   0:10.16 consul
$ sudo kill -USR1 `pgrep consul`
$ sudo journalctl -u consul | grep 'consul.runtime.alloc_bytes' | tail -n 1
Oct 01 16:20:06 consul-1 consul[3936]: [2018-10-01 16:19:50 +0000 UTC][G] 'consul.runtime.alloc_bytes': 5423240.000
$ sudo journalctl -u consul | grep 'consul.runtime.sys_bytes' | tail -n 1
Oct 01 16:20:06 consul-1 consul[3936]: [2018-10-01 16:19:50 +0000 UTC][G] 'consul.runtime.sys_bytes': 13052152.000
```

Here, `consul.runtime.sys_bytes`, is roughly a third of the virtual memory used by the Consul process. What does that 13052152 number mean exactly?

### Consul info for both Client and Server

```
$ consul info
agent:
        check_monitors = 0
        check_ttls = 0
        checks = 0
        services = 0
build:
        prerelease =
        revision = 9a494b5f
        version = 1.0.6
consul:
        bootstrap = false
        known_datacenters = 1
        leader = false
        leader_addr = 192.168.50.3:8300
        server = true
raft:
        applied_index = 928
        commit_index = 928
        fsm_pending = 0
        last_contact = 10.898525ms
        last_log_index = 928
        last_log_term = 2
        last_snapshot_index = 0
        last_snapshot_term = 0
        latest_configuration = [{Suffrage:Voter ID:2f1a09ad-25ff-8f6b-f15f-175bd89604fe Address:192.168.50.4:8300} {Suffrage:Voter ID:d7199834-fd33-eead-b5c9-3fa655b8e4b4 Address:192.168.50.3:8300} {Suffrage:Voter ID:6de32749-14a5-5bc5-ebb8-b6cc5a2aa326 Address:192.168.50.2:8300}]
        latest_configuration_index = 1
        num_peers = 2
        protocol_version = 3
        protocol_version_max = 3
        protocol_version_min = 0
        snapshot_version_max = 1
        snapshot_version_min = 0
        state = Follower
        term = 2
runtime:
        arch = amd64
        cpu_count = 1
        goroutines = 84
        max_procs = 1
        os = linux
        version = go1.9.3
serf_lan:
        coordinate_resets = 0
        encrypted = false
        event_queue = 0
        event_time = 2
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 6
        members = 5
        query_queue = 0
        query_time = 1
serf_wan:
        coordinate_resets = 0
        encrypted = false
        event_queue = 0
        event_time = 1
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 6
        members = 3
        query_queue = 0
        query_time = 1
```

### Operating system and Environment details

```
$ cat /proc/cpuinfo
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 70
model name      : Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
stepping        : 1
cpu MHz         : 2494.226
cache size      : 6144 KB
physical id     : 0
siblings        : 1
core id         : 0
cpu cores       : 1
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 13
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 syscall nx rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc eagerfpu pni pclmulqdq monitor ssse3 cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx rdrand hypervisor lahf_lm abm fsgsbase avx2 invpcid
bogomips        : 4988.45
clflush size    : 64
cache_alignment : 64
address sizes   : 39 bits physical, 48 bits virtual
power management:

$ cat /proc/meminfo
MemTotal:        1015500 kB
MemFree:          441788 kB
MemAvailable:     739852 kB
Buffers:            2108 kB
Cached:           412632 kB
SwapCached:            0 kB
Active:           283904 kB
Inactive:         189332 kB
Active(anon):      59108 kB
Inactive(anon):     6284 kB
Active(file):     224796 kB
Inactive(file):   183048 kB
Unevictable:           0 kB
Mlocked:               0 kB
SwapTotal:       2097148 kB
SwapFree:        2097148 kB
Dirty:                 0 kB
Writeback:             0 kB
AnonPages:         58528 kB
Mapped:            35500 kB
Shmem:              6896 kB
Slab:              62296 kB
SReclaimable:      35992 kB
SUnreclaim:        26304 kB
KernelStack:        2272 kB
PageTables:         5044 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:     2604896 kB
Committed_AS:     328204 kB
VmallocTotal:   34359738367 kB
VmallocUsed:       12096 kB
VmallocChunk:   34359718056 kB
HardwareCorrupted:     0 kB
AnonHugePages:     14336 kB
CmaTotal:              0 kB
CmaFree:               0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
DirectMap4k:       51136 kB
DirectMap2M:      997376 kB
$ uname -a
Linux consul-1 3.10.0-862.6.3.el7.x86_64 #1 SMP Tue Jun 26 16:32:21 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
$ cat /etc/centos-release
CentOS Linux release 7.5.1804 (Core)
```

Thanks for your help!

**Comment by pearkes** [comment]
Hey @jrab89. Consul is written in Go and uses a library called go-metrics to generate this telemetry from the Go runtime package.

You can reference these two links to see how these map back to values as defined in the runtime package in Go:

- https://github.com/armon/go-metrics/blob/58588f401c2cc130a7308a52ca3bc6c0a76db04b/metrics.go#L244-L250
- https://golang.org/pkg/runtime/#MemStats

I'm going to tag this as a docs issue as we could go into more depth based on the above for the runtime stats.

**Comment by shaqq** [comment]
@jrab89

I've done a bit of googling around on this, and here's what I found:

TL;DR - it's expected for there to be a discrepancy between what the OS says (via `top`) and what the program says (`MemStats`)

Some links!

- https://golang.org/doc/faq#Why_does_my_Go_process_use_so_much_virtual_memory
- https://stackoverflow.com/a/24865558
- https://golang.org/pkg/runtime/#MemStats (@pearkes linked this earlier)
- https://github.com/golang/go/issues/15849 (an interesting GH issue on how they updated the golang docs)

I don't have a complete answer for you since it's not possible for me to reproduce your scenario exactly and true up the numbers without your machine.

@pearkes regarding updating the docs - since this is more about Go rather than Consul, it might be sufficient to update the docs to say that those metrics are reflecting numbers from MemStats, and then point a link to the MemStats docs. What do you think? I'm happy to push up a PR!
